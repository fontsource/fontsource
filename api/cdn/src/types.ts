type CFRouterContext = [env: Env, ctx: ExecutionContext];

interface CSSFilename {
	isIndex?: boolean;
	isVariable?: boolean;

	style?: string;
	subset?: string;
	weight?: number;
	axes?: string;
}

export type { CFRouterContext, CSSFilename };
