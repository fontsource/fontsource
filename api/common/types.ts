type FontVariants = Record<
	string,
	Record<
		string,
		Record<
			string,
			{
				url: {
					woff2: string;
					woff: string;
					ttf?: string;
					otf?: string;
				};
			}
		>
	>
>;

interface IDResponse {
	id: string;
	family: string;
	subsets: string[];
	weights: number[];
	styles: string[];
	defSubset: string;
	variable: boolean;
	lastModified: string;
	category: string;
	license: string;
	type: string;
	unicodeRange: Record<string, string>;
	variants: FontVariants;
}

export type { FontVariants, IDResponse };
