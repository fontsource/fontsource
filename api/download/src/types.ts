type CFRouterContext = [env: Env, ctx: ExecutionContext];
interface FileGenerator {
	id: string;
	subsets: string[];
	weights: number[];
	styles: string[];
}

export type { CFRouterContext, FileGenerator };
