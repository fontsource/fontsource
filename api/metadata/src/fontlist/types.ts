import { type IDResponse } from 'common-api/types';

import type { FontMetadata } from '../types';

const fontlistQueries = [
	'family',
	'subsets',
	'weights',
	'styles',
	'variable',
	'lastModified',
	'category',
	'version',
	'type',
] as const;
type FontlistQueries = (typeof fontlistQueries)[number] & keyof FontMetadata;
const isFontlistQuery = (query: string): query is FontlistQueries =>
	fontlistQueries.includes(query as FontlistQueries);

type Fontlist = Record<string, string | string[] | number[] | boolean>;

type MetadataList = Record<string, IDResponse>;

export { fontlistQueries, isFontlistQuery };
export type { Fontlist, FontlistQueries, MetadataList };
