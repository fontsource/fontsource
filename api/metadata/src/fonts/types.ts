interface ArrayMetadataItem {
	id: string;
	family: string;
	subsets: string[];
	weights: number[];
	styles: string[];
	defSubset: string;
	variable: boolean;
	lastModified: string;
	category: string;
	license: string;
	type: string;
}

type ArrayMetadata = ArrayMetadataItem[];

const fontsQueries = [
	'id',
	'family',
	'subsets',
	'weights',
	'styles',
	'defSubset',
	'variable',
	'lastModified',
	'category',
	'license',
	'type',
] as const;
type FontsQueries = (typeof fontsQueries)[number] & keyof ArrayMetadataItem;
const isFontsQueries = (x: string): x is FontsQueries =>
	fontsQueries.includes(x as FontsQueries);

interface FontVariants {
	[weight: string]: {
		[style: string]: {
			[subset: string]: {
				url: {
					woff2: string;
					woff: string;
					ttf?: string;
					otf?: string;
				};
			};
		};
	};
}

interface IDResponse {
	id: string;
	family: string;
	subsets: string[];
	weights: number[];
	styles: string[];
	defSubset: string;
	variable: boolean;
	lastModified: string;
	category: string;
	license: string;
	type: string;
	unicodeRange: Record<string, string>;
	variants: FontVariants;
}

export type {
	ArrayMetadata,
	ArrayMetadataItem,
	FontsQueries,
	FontVariants,
	IDResponse,
};
export { isFontsQueries };
