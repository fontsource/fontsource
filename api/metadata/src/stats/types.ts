interface StatsResponse {
	npmDownloadTotal: number;
	npmDownloadMonthly: number;
	jsDelivrHitsTotal: number;
	jsDelivrHitsMonthly: number;
}

interface StatsResponseAll {
	total: StatsResponse;
	static: StatsResponse;
	variable?: StatsResponse;
}

type StatsResponseAllRecord = Record<string, StatsResponseAll>;

export type { StatsResponse, StatsResponseAll, StatsResponseAllRecord };
