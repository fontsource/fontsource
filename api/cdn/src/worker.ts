import { error } from 'itty-router';

import router, { corsify } from './router';

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext,
	): Promise<Response> {
		return await router
			.handle(request, env, ctx)
			.catch((error_) => {
				console.error(error_);
				const resp = error(error_);
				// Our jsDelivr proxy will fallback to one hour if no cache-control is set
				resp.headers.set('Cache-Control', 'max-age=0');
				return resp;
			})
			.then(corsify);
	},
};
