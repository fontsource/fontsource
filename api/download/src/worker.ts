import router from './router';
import { error } from 'itty-router';

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext
	): Promise<Response> {
		return router.handle(request, env, ctx).catch((err) => {
			console.error(err);
			return error(err);
		});
	},
};
