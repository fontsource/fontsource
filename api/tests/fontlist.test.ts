import { afterAll, beforeAll, beforeEach, expect, it, vi } from 'vitest';
const describe = setupMiniflareIsolatedStorage();
import worker from '../src/worker';
import { fontlistQueries } from '../src/fontlist/types';
import { mockMetadata } from './helpers';

describe('fontlist worker', () => {
	const env = getMiniflareBindings() as Env;
	const ctx = new ExecutionContext();

	beforeEach(async () => {
		mockMetadata();
	});

	it('should return type as default request', async () => {
		const request = new Request('http://localhost:8787/fontlist');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toEqual({
			abel: 'google',
			'advent-pro': 'google',
			'material-icons': 'google',
			'material-symbols-outlined': 'google',
			yakuhanjp: 'other',
		});
	});

	const queryParams = fontlistQueries;
	for (const query of queryParams) {
		it(`should return return value of ${query} query string`, async () => {
			const request = new Request(`http://localhost:8787/fontlist?${query}`);
			const response = await worker.fetch(request, env, ctx);
			expect(response.status).toBe(200);

			const body = await response.json();
			expect(body).toMatchSnapshot();
		});
	}

	it('should return 400 for invalid query string', async () => {
		const request = new Request(`http://localhost:8787/fontlist?invalid`);
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(400);
	});

	it('should return 400 for multiple query strings', async () => {
		const request = new Request(`http://localhost:8787/fontlist?type&category`);
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(400);
	});
});
