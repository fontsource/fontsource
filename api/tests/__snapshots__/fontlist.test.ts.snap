// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fontlist worker > should return return value of category query string 1`] = `
{
  "abel": "sans-serif",
  "advent-pro": "sans-serif",
  "material-icons": "icons",
  "material-symbols-outlined": "icons",
  "yakuhanjp": "other",
}
`;

exports[`fontlist worker > should return return value of family query string 1`] = `
{
  "abel": "Abel",
  "advent-pro": "Advent Pro",
  "material-icons": "Material Icons",
  "material-symbols-outlined": "Material Symbols Outlined",
  "yakuhanjp": "YakuHanJP",
}
`;

exports[`fontlist worker > should return return value of lastModified query string 1`] = `
{
  "abel": "2022-09-22",
  "advent-pro": "2023-03-21",
  "material-icons": "2023-03-08",
  "material-symbols-outlined": "2023-04-27",
  "yakuhanjp": "2023-05-15",
}
`;

exports[`fontlist worker > should return return value of styles query string 1`] = `
{
  "abel": [
    "normal",
  ],
  "advent-pro": [
    "italic",
    "normal",
  ],
  "material-icons": [
    "normal",
  ],
  "material-symbols-outlined": [
    "normal",
  ],
  "yakuhanjp": [
    "normal",
  ],
}
`;

exports[`fontlist worker > should return return value of subsets query string 1`] = `
{
  "abel": [
    "latin",
  ],
  "advent-pro": [
    "cyrillic",
    "cyrillic-ext",
    "greek",
    "latin",
    "latin-ext",
  ],
  "material-icons": [
    "latin",
  ],
  "material-symbols-outlined": [
    "latin",
  ],
  "yakuhanjp": [
    "latin",
  ],
}
`;

exports[`fontlist worker > should return return value of type query string 1`] = `
{
  "abel": "google",
  "advent-pro": "google",
  "material-icons": "google",
  "material-symbols-outlined": "google",
  "yakuhanjp": "other",
}
`;

exports[`fontlist worker > should return return value of variable query string 1`] = `
{
  "abel": false,
  "advent-pro": true,
  "material-icons": false,
  "material-symbols-outlined": true,
  "yakuhanjp": false,
}
`;

exports[`fontlist worker > should return return value of version query string 1`] = `
{
  "abel": "v18",
  "advent-pro": "v20",
  "material-icons": "v140",
  "material-symbols-outlined": "v108",
  "yakuhanjp": "v3.3.1",
}
`;

exports[`fontlist worker > should return return value of weights query string 1`] = `
{
  "abel": [
    400,
  ],
  "advent-pro": [
    100,
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
  ],
  "material-icons": [
    400,
  ],
  "material-symbols-outlined": [
    100,
    200,
    300,
    400,
    500,
    600,
    700,
  ],
  "yakuhanjp": [
    100,
    200,
    300,
    400,
    500,
    700,
    900,
  ],
}
`;
